### PR TITLE
Skip bad encrypted form instances during export. Fixes #10

### LINKED ISSUE #10: Briefcase pulls incomplete instances of encrypted forms, causes error when decrypting
<a href="https://github.com/mitchellsundt"></img></a> **Issue by [mitchellsundt](https://github.com/mitchellsundt)**
_Thursday Jul 09, 2015 at 19:01 GMT_
_Originally opened as https://github.com/opendatakit/opendatakit/issues/671 (1 comment(s))_

----

Originally reported on Google Code with ID 670

```
What steps will reproduce the problem?
1.Use briefcase to pull instances from an encrypted form off of a device
2. One or more instances is incomplete
3. Decryption/exporting is interrupted and fails with the following error: Error decrypting
submission .... Cause:
org.opendatakit.briefcase.model.ParsingException: Missing
base64EncryptedKey element in encrypted form.


What is the expected output? What do you see instead?
Expected output is to export/decrypt without interruption, somehow skipping over the
incomplete instances that lack the encrypted key element.  Instead, the export fails
and it is necessary to manually remove incomplete instance from briefcase storage before
completing decryption.

What version of the product are you using? On what operating system?
Using ODK Collect v1.2 rev 1013, ODK Briefcase v1.0.2.1, XLSForm on Android 2.2

Please provide any additional information below.
```

Reported by `abarnett5989` on 2012-08-22 17:05:53

### PR DESCRIPTION
**Pull request by icemc**
@yanokwa @lognaturel concerning what was discussed [here](https://github.com/opendatakit/briefcase/pull/84#issuecomment-292670209) during pulling of form instances if one of the instances could not be downloaded then it is skipped the pulling continues but at the end the whole process is marked as FAILED! even though the process completed and was not aborted. For my opinion i think such events should be marked as "SUCCESS WITH SOME ERRORS". So i went forward to implement  such an event during CSV export. If that looks nice to you guys I could extend it to the pull process. Here is a screen shot
successerrors
Note that if all the files were skipped then the export process is marked as "FAILED".
This PR fixes #10 
Will be absent for a while. But if this functionality sounds great i would like to extend it in the push and pull process too. Cheers

### CODE REVIEW

File: src/org/opendatakit/briefcase/util/ExportToCsv.java [hunk 1]
```diff
@@ -851,21 +854,18 @@ private boolean processInstance(File instanceDir) {
...
-          return false;
-        } catch (CryptoException e) {
-          e.printStackTrace();
+        } catch (ParsingException | CryptoException | FileSystemException e) {
+          //Was unable to parse file or decrypt file or a file system error occurred
+          //Hence skip this instance
           EventBus.publish(new ExportProgressEvent("Error decrypting submission "
-              + instanceDir.getName() + " Cause: " + e.toString()));
-          return false;
+                  + instanceDir.getName() + " Cause: " + e.toString() + " skipping...."));
```
Comment (yanokwa) [inline src/org/opendatakit/briefcase/util/ExportToCsv.java]: These strings can probably be combined with the lines below.

File: src/org/opendatakit/briefcase/util/ExportToCsv.java [hunk 2]
```diff
@@ -960,4 +960,14 @@ private boolean processInstance(File instanceDir) {
...
   }
+
+  @Override
+  public FilesSkipped totalFilesSkipped() {
+    //Determine if all files where skipped or just some
+    //Note that if totalInstances = 0 then no files were skipped
+    if (totalIntances == 0) {
+      return FilesSkipped.NONE;
+    }
+    return totalFilesSkipped == totalIntances ? FilesSkipped.ALL : FilesSkipped.SOME;
```
Comment (yanokwa) [inline src/org/opendatakit/briefcase/util/ExportToCsv.java]: I'd avoid ternary operations.

File: src/org/opendatakit/briefcase/ui/ExportPanel.java [hunk 3]
```diff
@@ -577,6 +578,12 @@ public void successfulCompletion(ExportSucceededEvent event) {
     lblExporting.setText("SUCCEEDED!");
     setActiveExportState(false);
   }
+  @EventSubscriber(eventClass = ExportSucceededWithErrorsEvent.class)
+  public void successfulCompletionWithErrors(ExportSucceededWithErrorsEvent event) {
+    exportStatusList.append("\n").append("SUCCEEDED WITH ERRORS!");
+    lblExporting.setText("SUCCEEDED WITH SOME ERRORS SEE DETAILS!");
```
Comment (yanokwa) [inline src/org/opendatakit/briefcase/ui/ExportPanel.java]: Let's make this less CAPS. `Succeeded, but with errors. See details!`

File: src/org/opendatakit/briefcase/ui/ExportPanel.java [hunk 4]
```diff
@@ -577,6 +578,12 @@ public void successfulCompletion(ExportSucceededEvent event) {
     lblExporting.setText("SUCCEEDED!");
     setActiveExportState(false);
   }
+  @EventSubscriber(eventClass = ExportSucceededWithErrorsEvent.class)
+  public void successfulCompletionWithErrors(ExportSucceededWithErrorsEvent event) {
+    exportStatusList.append("\n").append("SUCCEEDED WITH ERRORS!");
```
Comment (yanokwa) [inline src/org/opendatakit/briefcase/ui/ExportPanel.java]: Let's make this less CAPS. `Succeeded, but with errors!`

**Comment by yanokwa** [comment]
I'm pretty happy with the functionality. @lognaturel @nribeka Can you review the code when you get a moment?

@icemc Great work on this PR! When you get back, can you please file a detailed issue for the pull process?

**Comment by yanokwa** [comment]
@icemc I usually don't push code into someone's branch, so very sorry about that. I had made some changes to master that caused a number of merge conflicts. Since I caused the problem, I figured I should fix it. Thanks again for this PR and I'm looking forward to PR to the changes to Pull!

**Comment by icemc** [comment]
@yanokwa no problem I understand. I'm glad you liked the functionality. Will file an issue and submit a PR about adding this to the pull process ASAP. Cheers

**Pull request merged**